validator_client/src/validator_duty.rs

```diff
@@ -53,58 +55,103 @@ impl ValidatorDuty {
         beacon_node: &BeaconNodeHttpClient,
         current_epoch: Epoch,
         request_epoch: Epoch,
-        pubkey: PublicKey,
-    ) -> Result<ValidatorDuty, String> {
-        let pubkey_bytes = PublicKeyBytes::from(&pubkey);
+        pubkeys: &[(PublicKey, Option<u64>)],
+        log: &Logger,
+    ) -> Result<Vec<ValidatorDuty>, String> {
+        let mut duties = Vec::with_capacity(pubkeys.len());
+        let mut query_indices = Vec::with_capacity(pubkeys.len());
+        let mut query_pubkeys = Vec::with_capacity(pubkeys.len());
 
-        let validator_index = if let Some(index) = beacon_node
-            .get_beacon_states_validator_id(
-                StateId::Head,
-                &ValidatorId::PublicKey(pubkey_bytes.clone()),
-            )
-            .await
-            .map_err(|e| format!("Failed to get validator index: {}", e))?
-            .map(|body| body.data.index)
-        {
-            index
+        for (pubkey, index_opt) in pubkeys {
+            if let Some(index) = index_opt {
+                // If we know the index already, include it in the duties query
+                query_indices.push(*index);
+                query_pubkeys.push(pubkey.clone());
+            } else {
+                // Query for any pubkeys we don't know the index for in the current epoch.
+                let pubkey_bytes = PublicKeyBytes::from(pubkey);
+                if let Some(index) = beacon_node
+                    .get_beacon_states_validator_id(
+                        StateId::Head,
+                        &ValidatorId::PublicKey(pubkey_bytes.clone()),
+                    )
+                    .await
+                    .map_err(|e| format!("Failed to get validator index: {}", e))?
+                    .map(|body| body.data.index)
+                {
+                    query_indices.push(index);
+                    query_pubkeys.push(pubkey.clone());
+                } else {
+                    // If we still don't know the index, add an empty duty
+                    duties.push(Self::no_duties(pubkey.clone(), None));
```

Comment:
I think we lose the 1:1 ordering between request and response here.

I had a go at maintaining that ordering here: https://github.com/sigp/lighthouse/commit/107fd2e1657f69a71b8fe9fad23b4a1805756f57

Reply:
I cherry-picked this commit, but removed lines 130-140 because all entries in the `attester_data_map` are (`attester_data.validator_index` : `attester_data`), so it's it's a comparison of `attester_data.validator_index` to itself

Reply:
> so it's it's a comparison of attester_data.validator_index to itself

Good point!